Sprint 2/step 2

yaXXrat opened the pull request.

xxrat committed:
First step: initialState, structure

yaXXrat committed:
1. start point use Redux for order creation. 2. service folder structure changed 3. Code clean-up

yaXXrat committed:
Erase order on close order details info window

yaXXrat committed:
1. Add ingredient to constructor by click TEMPORARY 2. Remove ingredient from constructor by click on garbage icon

xxrat committed:
Modals into Redux: Error, selectIngredient

yaXXrat committed:
Show corresponded modal window when requested objects (order and selected ingredient) are exist

yaXXrat committed:
Code clean-up: 1. Move error to separate reducer 2. Rename reducers

yaXXrat committed:
1. Burger Constructor refactoring: new component for bun 2. Remove unused context 3. Code clean-up

yaXXrat committed:
Switch tabs during scroll

xxrat committed:
burgerConstructor DND

yaXXrat committed:
lastIndex for order state was added

xxrat committed:
burgerConstructor DND bug fixed

yaXXrat committed:
1. Drag'n'Drop for burger ingredients 2. Code clean-up

xxrat committed:
Waiting Modal, REQUEST, SUCCESS, ERROR actions

yaXXrat committed:
Middleware for fetching operations

xxrat committed:
bun in initial order, minor changes, code cleanup

andrepolischuk commented on src/components/app/app.js:
```diff
@@ -8,151 +8,76 @@ import Modal from '../modal/modal';
 
-  const createOrder = (createdOrder) => {
-    setOrder(createdOrder);
+  const hideDisplayWaiting = () => {
```
можно сделать лучше: пустую функцию лучше удалить, чтобы не замусоривать кодовую базу

andrepolischuk reviewed (commented):
доброе утро, хорошая работа
- все работает как ожидается
- реализовано переключение навигации по ингредиентам
- при перетаскивании ингредиента в конструктор добавляется кастомный уникальный идентификатор

но нужно немного исправить
- по умолчанию в конструкторе должно быть пусто, сейчас сразу добавляется краторная булка
- нужно убрать лишние таймауты, чтобы не создавать пользователю дополнительное ожидание в секунду
- каждый редьюсер должен решать свою задачу, поэтому нужно разделить редьюсеры и экшны, чтобы в итоге было 4 отдельных редьюсера:
  - загрузка и хранение ингредиентов
  - работа конструктора
  - управление модальным окном ингредиента
  - оформление заказа
- если при выполнении запроса к api произошла ошибка, то содержимое соответствующего хранилища должно приводится к начальному состоянию
- переменные которые не меняются, должны быть объявлены через const
- в Modal нужно проверять на тип пропса onClose, например что typeof onClose это function

и можно сделать еще лучше
- пустые и неиспользуемые функции и классы стилей лучше удалить, чтобы не замусоривать кодовую базу
- лучше в константу выносить только сам базовый адрес сервера https://norma.nomoreparties.space/api/, а endpoint добавлять отдельно в функции запроса
- лучше перенести getIngredients/createOrder в соответствующие файлы экшнов, чтобы было понятнее к чему они относятся
- лучше не дублировать функционал и оставить в хранилище только errorMessage, если errorMessage есть - показываем ошибку в интерфейсе
- лучше не дублировать функционал и оставить в хранилище только selectedIngredient и если он есть, показывать модалку
- лучше убрать orderBunSelected из хранилища, потому что он по сути дублирует orderBun, чтобы показать булку в конструкторе, достаточно одного поля
- чтобы уменьшить количество срабатываний обработчика скролла по списку ингредиентов и расчетов имеет смысл обернуть его в throttle

работа может быть принята только после исправления всех критических замечаний помеченных `нужно исправить`

удачи

xxrat committed:
Review first part fixed:
1. No initial bun
2. No timeouts
3. New reducer & action for modal ingredient
4. Error => initial state
5. Right onClose in Modal
etc.

yaXXrat committed:
Fixes according reviewer comments

yaXXrat committed:
Another part fixes

xxrat committed:
Final cleanup: no more boolean flags etc

xxrat committed:
Fixed small bug of error catching

xxrat committed:
Another fixed small bug of error catching

andrepolischuk commented on src/components/burger-constructor/burger-constructor.js:
```diff
@@ -8,27 +8,19 @@ import { UPDATE_ORDER, SET_ORDER_INGREDIENT } from '../../services/actions/order
-            price: totalPrice
-        }
-        dispatch(createOrder(ingredientsIDs,newOrder));
+        dispatch(createOrder(ingredientsIDs,totalPrice));
```
так даже лучше, что передаются только нужные данные 👍

andrepolischuk commented on src/components/burger-ingredients/burger-ingredients.js:
```diff
@@ -50,6 +64,7 @@ const BurgerIngredients = () => {
           .sort((a, b) => a[1] - b[1])[0][0];
         setCurrentIngredientsType(activeTab);    
     }
+    const throttledUpdateTab = throttle(updateTab, 250);
```
нужно исправить: поскольку throttle имеет внутри себя счетчик для отсечения лишних срабатываний, нужно функцию throttle положить в отдельный модуль, например в utils, а функции updateTab обернуть в useCallback, чтобы функция и счетчик внутри нее не создавались заново на каждый ререндер

```js
import throttle from '../../utils/throttle'

...

const updateTab = useCallback(() => {
  ...
}, [])

const throttledUpdateTab = useCallback(throttle(updateTab, 250), [updateTab])
```

yaXXrat replied:
@andrepolischuk Увы, ругается:  
Line 56:7: React Hook useCallback has missing dependencies: 'groupContainerTop' and 'tabsRef'. Either include them or remove the dependency array  react-hooks/exhaustive-deps
Line 58:32:  React Hook useCallback received a function whose dependencies are unknown. Pass an inline function instead

Или нам отключить elint на этих строках?

yaXXrat replied:
@andrepolischuk А [вот тут](https://habr.com/ru/post/529950/) утверждается, что «основная идея не в улучшении перформанса в конкретном компоненте, а использование useCallback выгодно только в случае передачи функции как props».

andrepolischuk replied:
> Line 56:7: React Hook useCallback has missing dependencies: 'groupContainerTop' and 'tabsRef'. Either include them or remove the dependency array react-hooks/exhaustive-deps

тут да, надо добавить - я пустой массив указал для примера

> Line 58:32: React Hook useCallback received a function whose dependencies are unknown. Pass an inline function instead

да, действительно. чтобы зависимости не терялись можно использовать useMemo вместо useCallback https://kyleshevlin.com/debounce-and-throttle-callbacks-with-react-hooks

```js
const throttledUpdateTab = useMemo(() => throttle(updateTab, 250), [updateTab])
```

> А вот тут утверждается, что «основная идея не в улучшении перформанса в конкретном компоненте, а использование useCallback выгодно только в случае передачи функции как props».

в данном кейсе с throttle, для улучшения перформанса как раз throttle, а useCallback нужен чтобы не дергать throttle в каждом рендере для сохранения счетчика вызовов внутри

yaXXrat replied:
@andrepolischuk Не уверен в твоем подходе, ну да ладно.
Проблемы это не решает. Добавил зависимости, но дальше предлагается использовать useMemo для tabsRef, чего я не могу сделать: этот объект инициализируется useRef.

andrepolischuk reviewed (commented):
добрый вечер, в целом все отлично, внесены практически все правки и даже то что можно было сделать лучше

но еще остался один недочет - нужно все таки разделить reducers/order и экшны к нему на 2 отдельных редьюсера, чтобы каждый из них выполнял одну задачу: один - хранение ингредиентов конструктора а второй - хранение данных оформленного заказа

и появился новый - поскольку throttle имеет внутри себя счетчик для отсечения лишних срабатываний, нужно функцию throttle положить в отдельный модуль, например в utils, а функции updateTab обернуть в useCallback, чтобы функция и счетчик внутри нее не создавались заново на каждый ререндер

исправьте оставшиеся недоработки и работа будет принята

удачи

yaXXrat committed:
According to reviewer comment: Split order reducer to order and constructor reducers

yaXXrat committed:
Erase constructor on order creation complete

yaXXrat committed:
Move throttle function to separate place

xxrat committed:
Change objects to initial 'undefined', try to use Callback for throttle.

xxrat committed:
Try to use reviewer's ideas.

xxrat committed:
Fixed using useMemo: createRef instead of useRef

yaXXrat committed:
Code clean-up

andrepolischuk commented on src/components/burger-ingredients/burger-ingredients.js:
```diff
@@ -42,29 +46,18 @@ const BurgerIngredients = () => {
+        setCurrentIngredientsType(activeTab);
+    },[groupContainerTop, tabsRef]);
+
+    const throttledUpdateTab = useMemo(() => throttle(updateTab, 250), [updateTab])
```
мемоизация сделана отлично, и для рефов тоже 👍

andrepolischuk commented on src/services/reducers/index.js:
```diff
@@ -1,12 +1,14 @@
 import { burgerIngredientsReducer } from "./ingredients";
 import { burgerIngredientReducer } from "./ingredient";
 
 export const rootReducer = combineReducers({
```
стор отлично поделен по функциям 👍

andrepolischuk reviewed (approved):
отлично, все недочеты исправлены и сделаны моменты которые можно было улучшить, ваша работа принята 🎉 

успехов в дальнейшем обучении
